File: frontend/src/types/cook.ts
// TypeScript interfaces matching backend models

export type MeatCategory = 'beef' | 'pork' | 'poultry' | 'lamb';

export type CutType =
  | 'brisket' | 'pork_butt' | 'pork_ribs' | 'beef_ribs'
  | 'chicken_whole' | 'turkey_breast' | 'leg_of_lamb';

export type CookState =
  | 'setup' | 'preheat' | 'early_cook' | 'pre_stall' | 'stall'
  | 'post_stall' | 'approaching_target' | 'rest' | 'done';

export type ConfidenceTier = 'high' | 'moderate' | 'low' | 'very_low';
export type EquipmentType = 'offset' | 'pellet' | 'kamado' | 'wsm' | 'custom';
export type WrapType = 'none' | 'foil' | 'butcher_paper' | 'foil_boat';
export type QualityRating = 'excellent' | 'good' | 'fair' | 'poor';

// Request types
export interface CookSetupRequest {
  meat_category: MeatCategory;
  cut_type: CutType;
  weight_lbs: number;
  thickness_inches: number;
  equipment_type: EquipmentType;
  smoker_temp_f: number;
  target_temp_f: number;
  dinner_time?: string;
  altitude_ft: number;
  latitude?: number;
  longitude?: number;
}

export interface ProbeReadingRequest {
  temp_f: number;
  smoker_temp_f?: number;
}

export interface WrapRequest {
  wrap_type: WrapType;
}

export interface FinishCookRequest {
  quality_rating?: QualityRating;
  quality_notes?: string;
}

// Response types
export interface PredictionResponse {
  p10_minutes: number;
  p50_minutes: number;
  p90_minutes: number;
  p10_time?: string;
  p50_time?: string;
  p90_time?: string;
  confidence: ConfidenceTier;
  current_state: CookState;
  stall_probability: number;
  readings_count: number;
}

export interface StateResponse {
  current_state: CookState;
  confidence: ConfidenceTier;
  stall_active: boolean;
  stall_duration_minutes: number;
  wrap_type: WrapType;
  readings_count: number;
  elapsed_minutes: number;
}

export interface BackwardPlanResponse {
  dinner_time: string;
  fire_start_time: string;
  meat_on_time: string;
  estimated_cook_minutes_p90: number;
  rest_minutes: number;
  preheat_minutes: number;
}

export interface CookSetupResponse {
  session_id: string;
  prediction: PredictionResponse;
  backward_plan?: BackwardPlanResponse;
}

export interface ReadingResponse {
  elapsed_minutes: number;
  prediction: PredictionResponse;
  state: StateResponse;
}

export interface WrapResponse {
  wrap_type: WrapType;
  prediction: PredictionResponse;
  message: string;
}

export interface ReportResponse {
  session_id: string;
  total_cook_minutes: number;
  final_temp_f: number;
  stall_occurred: boolean;
  stall_duration_minutes: number;
  was_wrapped: boolean;
  wrap_type: WrapType;
  prediction_accuracy_minutes: number;
  readings_count: number;
  lid_opens_count: number;
  quality_rating?: QualityRating;
  quality_notes?: string;
  predicted_temps: number[];
  actual_temps: number[];
  residuals: number[];
}

export interface EquipmentPresetResponse {
  equipment_type: EquipmentType;
  name: string;
  temp_variance: number;
  recovery_time_min: number;
  temp_drop_on_lid_open: number;
  insulation_factor: number;
}

// Cut display info
export const CUT_INFO: Record<CutType, { label: string; category: MeatCategory; defaultTarget: number; defaultThickness: number }> = {
  brisket:        { label: 'Brisket',          category: 'beef',    defaultTarget: 203, defaultThickness: 5.0 },
  pork_butt:      { label: 'Pork Butt',        category: 'pork',    defaultTarget: 203, defaultThickness: 6.0 },
  pork_ribs:      { label: 'Pork Ribs',        category: 'pork',    defaultTarget: 195, defaultThickness: 1.5 },
  beef_ribs:      { label: 'Beef Ribs',         category: 'beef',    defaultTarget: 203, defaultThickness: 2.5 },
  chicken_whole:  { label: 'Whole Chicken',     category: 'poultry', defaultTarget: 165, defaultThickness: 4.0 },
  turkey_breast:  { label: 'Turkey Breast',     category: 'poultry', defaultTarget: 165, defaultThickness: 4.5 },
  leg_of_lamb:    { label: 'Leg of Lamb',       category: 'lamb',    defaultTarget: 145, defaultThickness: 4.0 },
};

File: backend/routers/cook.py
"""Cook session API endpoints."""

from datetime import datetime, timedelta
from fastapi import APIRouter, HTTPException

from ..models.schemas import (
    CookSetupRequest,
    CookSetupResponse,
    BackwardPlanResponse,
    FinishCookRequest,
    LidOpenRequest,
    PredictionResponse,
    ProbeReadingRequest,
    ReadingResponse,
    ReportResponse,
    StateResponse,
    WrapRequest,
    WrapResponse,
)
from ..services import cook_session_service as svc

router = APIRouter(prefix="/api/v1/cook", tags=["cook"])


def _prediction_to_response(pred, session_created_at=None) -> PredictionResponse:
    """Convert PredictionResult to PredictionResponse with absolute times."""
    base_time = session_created_at or datetime.utcnow()
    return PredictionResponse(
        p10_minutes=pred.p10_minutes,
        p50_minutes=pred.p50_minutes,
        p90_minutes=pred.p90_minutes,
        p10_time=base_time + timedelta(minutes=pred.p10_minutes),
        p50_time=base_time + timedelta(minutes=pred.p50_minutes),
        p90_time=base_time + timedelta(minutes=pred.p90_minutes),
        confidence=pred.confidence,
        current_state=pred.current_state,
        stall_probability=pred.stall_probability,
        readings_count=pred.readings_count,
    )


@router.post("/setup", response_model=CookSetupResponse)
async def setup_cook(request: CookSetupRequest):
    """Create a new cook session, fetch weather, run initial MC."""
    session, prediction, backward_plan = await svc.create_session(request)

    bp_response = None
    if backward_plan:
        bp_response = BackwardPlanResponse(
            dinner_time=backward_plan.dinner_time,
            fire_start_time=backward_plan.fire_start_time,
            meat_on_time=backward_plan.meat_on_time,
            estimated_cook_minutes_p90=backward_plan.estimated_cook_minutes_p90,
            rest_minutes=backward_plan.rest_minutes,
            preheat_minutes=backward_plan.preheat_minutes,
        )

    return CookSetupResponse(
        session_id=session.id,
        prediction=_prediction_to_response(prediction, session.created_at),
        backward_plan=bp_response,
    )


@router.post("/{session_id}/reading", response_model=ReadingResponse)
async def add_reading(session_id: str, request: ProbeReadingRequest):
    """Log a probe temperature reading."""
    try:
        session, prediction = await svc.add_reading(session_id, request)
    except ValueError:
        raise HTTPException(status_code=404, detail="Session not found")

    elapsed = session.readings[-1].elapsed_minutes if session.readings else 0.0

    return ReadingResponse(
        elapsed_minutes=elapsed,
        prediction=_prediction_to_response(prediction, session.created_at),
        state=StateResponse(
            current_state=session.current_state,
            confidence=session.confidence,
            stall_active=session.stall.in_stall,
            stall_duration_minutes=session.stall.stall_duration_minutes,
            wrap_type=session.wrap_type,
            readings_count=len(session.readings),
            elapsed_minutes=elapsed,
        ),
    )


@router.post("/{session_id}/lid-open")
async def lid_open(session_id: str, request: LidOpenRequest = LidOpenRequest()):
    """Log a lid-open event."""
    await svc.log_lid_open(session_id, request.duration_seconds)
    return {"status": "ok"}


@router.post("/{session_id}/wrap", response_model=WrapResponse)
async def apply_wrap(session_id: str, request: WrapRequest):
    """Log wrap decision, adjust model, re-run MC."""
    try:
        session, prediction, message = await svc.apply_wrap(session_id, request)
    except ValueError:
        raise HTTPException(status_code=404, detail="Session not found")

    return WrapResponse(
        wrap_type=request.wrap_type,
        prediction=_prediction_to_response(prediction, session.created_at),
        message=message,
    )


@router.get("/{session_id}/prediction", response_model=PredictionResponse)
async def get_prediction(session_id: str):
    """Get the latest cached prediction."""
    pred = await svc.get_prediction(session_id)
    if pred is None:
        raise HTTPException(status_code=404, detail="No prediction found")

    session = await svc.get_session(session_id)
    base_time = session.created_at if session else datetime.utcnow()
    return _prediction_to_response(pred, base_time)


@router.get("/{session_id}/state", response_model=StateResponse)
async def get_state(session_id: str):
    """Get current cook state and confidence."""
    session = await svc.get_session(session_id)
    if session is None:
        raise HTTPException(status_code=404, detail="Session not found")

    elapsed = session.readings[-1].elapsed_minutes if session.readings else 0.0
    return StateResponse(
        current_state=session.current_state,
        confidence=session.confidence,
        stall_active=session.stall.in_stall,
        stall_duration_minutes=session.stall.stall_duration_minutes,
        wrap_type=session.wrap_type,
        readings_count=len(session.readings),
        elapsed_minutes=elapsed,
    )


@router.post("/{session_id}/finish", response_model=ReportResponse)
async def finish_cook(session_id: str, request: FinishCookRequest = FinishCookRequest()):
    """End cook and compute report."""
    try:
        report = await svc.finish_cook(session_id, request)
    except ValueError:
        raise HTTPException(status_code=404, detail="Session not found")

    return ReportResponse(
        session_id=report.session_id,
        total_cook_minutes=report.total_cook_minutes,
        final_temp_f=report.final_temp_f,
        stall_occurred=report.stall_occurred,
        stall_duration_minutes=report.stall_duration_minutes,
        was_wrapped=report.was_wrapped,
        wrap_type=report.wrap_type,
        prediction_accuracy_minutes=report.prediction_accuracy_minutes,
        readings_count=report.readings_count,
        lid_opens_count=report.lid_opens_count,
        quality_rating=report.quality_rating,
        quality_notes=report.quality_notes,
        predicted_temps=report.predicted_temps,
        actual_temps=report.actual_temps,
        residuals=report.residuals,
    )


@router.get("/{session_id}/report", response_model=ReportResponse)
async def get_report(session_id: str):
    """Get post-cook report for a finished session."""
    report = await svc.get_report(session_id)
    if report is None:
        raise HTTPException(status_code=404, detail="Report not found")

    return ReportResponse(
        session_id=report.session_id,
        total_cook_minutes=report.total_cook_minutes,
        final_temp_f=report.final_temp_f,
        stall_occurred=report.stall_occurred,
        stall_duration_minutes=report.stall_duration_minutes,
        was_wrapped=report.was_wrapped,
        wrap_type=report.wrap_type,
        prediction_accuracy_minutes=report.prediction_accuracy_minutes,
        readings_count=report.readings_count,
        lid_opens_count=report.lid_opens_count,
        quality_rating=report.quality_rating,
        quality_notes=report.quality_notes,
        predicted_temps=report.predicted_temps,
        actual_temps=report.actual_temps,
        residuals=report.residuals,
    )
